services/storage-users/pkg/task/trash_bin.go

```diff
@@ -0,0 +1,120 @@
+package task
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	apiGateway "github.com/cs3org/go-cs3apis/cs3/gateway/v1beta1"
+	apiUser "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	apiRpc "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	apiProvider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/v2/pkg/errtypes"
+	"github.com/cs3org/reva/v2/pkg/utils"
+)
+
+// PurgeTrashBin can be used to purge space trash-bin's,
+// the provided executantID must have space access.
+// removeBefore specifies how long an item must be in the trash-bin to be deleted,
+// items that stay there for a shorter time are ignored and kept in place.
+func PurgeTrashBin(executantID *apiUser.UserId, deleteBefore time.Time, spaceType SpaceType, gwc apiGateway.GatewayAPIClient, machineAuthAPIKey string) error {
+	executantCtx, _, err := utils.Impersonate(executantID, gwc, machineAuthAPIKey)
+	if err != nil {
+		return err
+	}
+
+	listStorageSpacesResponse, err := gwc.ListStorageSpaces(executantCtx, &apiProvider.ListStorageSpacesRequest{
+		Filters: []*apiProvider.ListStorageSpacesRequest_Filter{
+			{
+				Type: apiProvider.ListStorageSpacesRequest_Filter_TYPE_SPACE_TYPE,
+				Term: &apiProvider.ListStorageSpacesRequest_Filter_SpaceType{
+					SpaceType: string(spaceType),
+				},
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, storageSpace := range listStorageSpacesResponse.StorageSpaces {
+		var (
+			err                   error
+			impersonationID       *apiUser.UserId
+			storageSpaceReference = &apiProvider.Reference{
+				ResourceId: storageSpace.GetRoot(),
+			}
+		)
+
+		switch storageSpace.GetSpaceType() {
+		case string(Personal):
+			impersonationID = storageSpace.GetOwner().GetId()
+		case string(Project):
+			opaqueGrants, ok := storageSpace.GetOpaque().GetMap()["grants"]
+			if !ok {
+				err = errors.New("no grants")
+				break
+			}
+
+			var permissionsMap map[string]*apiProvider.ResourcePermissions
+			err = json.Unmarshal(opaqueGrants.Value, &permissionsMap)
+			if err != nil {
+				break
+			}
+
+			for id, permissions := range permissionsMap {
+				if !permissions.Delete {
+					continue
+				}
+
+				impersonationID = &apiUser.UserId{
+					OpaqueId: id,
+				}
+				break
+			}
+		default:
+			continue
+		}
+
+		if impersonationID == nil {
```

Comment:
```suggestion
		if impersonationID == nil && err == nil {
```
Prolly better not to overwrite the error if it exists